### PR TITLE
distributed-lazy: delete unneeded temporaries after execution

### PR DESCRIPTION
**Pull request by matthiasdiener**
Add-on for #148.

Should help with https://github.com/illinois-ceesd/mirgecom/issues/599

### CODE REVIEW

File: pytato/distributed.py [hunk 1]
```diff
@@ -658,8 +655,20 @@ def wait_for_some_recvs() -> None:
                 and (set(partition.parts[pid].input_name_to_recv_node)
                     <= recv_names_completed)}
         for pid in ready_pids:
+
             exec_ready_part(partition.parts[pid])
 
+            partition_input_names_required = set()
+
+            for pid in pids_to_execute:
+                partition_input_names_required.add(
```
Comment (inducer) [inline pytato/distributed.py:664]:
```suggestion
                partition_input_names_required.update(
```

Comment (matthiasdiener) [reply pytato/distributed.py]: Done in 490b7ff

File: pytato/distributed.py [hunk 2]
```diff
@@ -609,6 +609,17 @@ def execute_distributed_partition(
...
+    # Keep a count on how often each input name is used,
+    # in order to be able to free them.
+    partition_input_names_refcount: Dict[str, int] = {}
+
+    for pid in pids_to_execute:
+        for name in partition.parts[pid].all_input_names():
+            if name in partition_input_names_refcount:
+                partition_input_names_refcount[name] += 1
+            else:
+                partition_input_names_refcount[name] = 1
```
Comment (inducer) [inline pytato/distributed.py:621]: This should be a `@memoize_method` on the partition. No need to recompute on every evaluation IMO. (Make sure to not modify the cached dictionary in place!)

File: pytato/distributed.py [hunk 3]
```diff
@@ -668,6 +681,10 @@ def wait_for_some_recvs() -> None:
     for send_req in send_requests:
         send_req.Wait()
 
+    if __debug__:
+        for _, v in partition_input_names_refcount.items():
+            assert v == 0
```
Comment (inducer) [inline pytato/distributed.py:686]: Maybe also assert that the values are gone from `context`?

File: pytato/distributed.py [hunk 4]
```diff
@@ -660,6 +668,11 @@ def wait_for_some_recvs() -> None:
         for pid in ready_pids:
             exec_ready_part(partition.parts[pid])
 
+            for p in partition.parts[pid].all_input_names():
```
Comment (inducer) [inline pytato/distributed.py:671]:
```suggestion
            part = partition.parts[pid]
            exec_ready_part(part)

            for p in part.all_input_names():
```

**Review by inducer — APPROVED** [review]
Thanks! LGTM subject to these minor fixes.

File: pytato/distributed.py (hunk 3 above)
Comment (matthiasdiener) [reply pytato/distributed.py]: Done in 54811252c64f993cb264b5b81f22a099a51bb079

File: pytato/distributed.py (hunk 2 above)
Comment (matthiasdiener) [reply pytato/distributed.py]: Done in 4b592fb

**Review by inducer — APPROVED** [review]
Thanks, LGTM! I'll aply the suggestion.

**Comment by inducer** [comment]
This should be able to land after https://github.com/inducer/pytools/pull/118 merges.

**Comment by matthiasdiener** [comment]
> This should be able to land after [inducer/pytools#118](https://github.com/inducer/pytools/pull/118) merges.

It will need a new pytools release+conda package I guess.

**Comment by inducer** [comment]
Yep, you're totally right. :facepalm: Doing that now.

**Pull request merged**